### PR TITLE
Introduce collectobjects(topology,:objtype) + convenience methods

### PR DESCRIPTION
**Pull request by j-fu**
Hi, I was struggling with obtaining different cache sizes from the example in the README (it didn't work for me), so I rolled my own and propose this addition to the higher level API.

Rationale:

Besides of num_physical_cores(), additional information is of
prevalent interest:

num_packages(): number of processor packages; Usually, each corresponds
                to one NUMA node with its localized memory. The existence of
                everal NUMA nodes tells about several lanes to memory
l*cache_sizes(): These are of pre-eminent interest with respect to performance
                on objects with larger memory footprint.

### CODE REVIEW

**Comment by nalimilan** [comment]
Any chance to get this reviewed and merged? The example README gives to find cache size doesn't work, so currently I don't know how to do this.

EDIT: I found a solution, see https://github.com/JuliaParallel/Hwloc.jl/pull/32. Though convenience methods would still be useful.

**Comment by carstenbauer** [comment]
Is there a particular reason why this hasn't been reviewed yet? (After all the PR is more than a year old.)

**Comment by j-fu** [comment]
yeah still would be cool to have this in the package...

**Comment by carstenbauer** [comment]
Actually, I have more ideas for improving this package and am motivated to implement them (I already filed a few PRs) but this PR made me question a bit whether supervision / maintenance of this package is still active. I hope to be wrong, of course :)

**Comment by nalimilan** [comment]
It seems that @eschnett still works on the package?

**Comment by eschnett** [comment]
@nalimilan Apologies for the delay. I'd be happy to add other maintainers to the package.

File: src/Hwloc.jl
Comment (carstenbauer) [inline src/Hwloc.jl]: 1) I would only accept `::Symbol`s. Should be fine.
2) Maybe simply implement it as
```Julia
collectobjects(t::Object, type_::Symbol) = collect(Iterators.filter(obj -> obj.type_ == type_, t))
```
? It's simpler, faster, and type inference doesn't fail:

```Julia
julia> @btime Hwloc.collectobjects2($topo, :L1Cache); # Iterators.filter
  2.398 μs (88 allocations: 5.83 KiB)

julia> @btime Hwloc.collectobjects($topo, :L1Cache); # PR
  8.419 μs (111 allocations: 4.97 KiB)
```

Iterators.filter:

```Julia
Variables
  #self#::Core.Const(Hwloc.collectobjects2)
  t::Hwloc.Object
  type_::Symbol
  #33::Hwloc.var"#33#34"{Symbol}

Body::Vector{Hwloc.Object}
1 ─ %1 = Base.getproperty(Hwloc.Iterators, :filter)::Core.Const(Base.Iterators.filter)
│   %2 = Hwloc.:(var"#33#34")::Core.Const(Hwloc.var"#33#34")
│   %3 = Core.typeof(type_)::Core.Const(Symbol)
│   %4 = Core.apply_type(%2, %3)::Core.Const(Hwloc.var"#33#34"{Symbol})
│        (#33 = %new(%4, type_))
│   %6 = #33::Hwloc.var"#33#34"{Symbol}
│   %7 = (%1)(%6, t)::Base.Iterators.Filter{Hwloc.var"#33#34"{Symbol}, Hwloc.Object}
│   %8 = Hwloc.collect(%7)::Vector{Hwloc.Object}
└──      return %8
```

PR:

```Julia
Variables
  #self#::Core.Const(Hwloc.collectobjects)
  t::Hwloc.Object
  type_::Symbol

Body::Any
1 ─ %1 = (#self#)(t, type_, Hwloc.nothing)::Any
└──      return %1
```

File: src/Hwloc.jl
Comment (carstenbauer) [inline src/Hwloc.jl]: It seems more natural to define the scalar `attributes(obj::Object)` and let the user use broadcasting to apply it to a collection of objects. What do you think?

File: src/Hwloc.jl
Comment (carstenbauer) [inline src/Hwloc.jl]: Probably better to just count the packages instead of collecting them into a vector and reading off the length:

```julia
num_packages() = count(obj->obj.type_ == :Package, Hwloc.topology_load())
```

Doesn't seem to be faster (because `Hwloc.topology_load()` dominates in both cases) but it's conceptually cleaner in my opinion.

File: src/Hwloc.jl
Comment (carstenbauer) [inline src/Hwloc.jl:482]: Similar to above, maybe better
```Julia
l1cache_sizes()=map(obj->obj.attr.size,Iterators.filter(obj->obj.type_ == :L1Cache, Hwloc.topology_load()))
```
instead of an explicit collect?

Personally, I'd also drop the underscore in the function names but that's a matter of taste I guess.

**Review by carstenbauer — CHANGES_REQUESTED** [review]

**Comment by carstenbauer** [comment]
@eschnett I would gladly help to improve and maintain this package.

**Comment by j-fu** [comment]
Hi @crstnbr: after  viewing the conversations I tend to agree with all of your proposals - they appear more Julian, probably I would do it differently now as well after gaining a year more  experience... So feel free to improve this, including the method names if they really hurt you...

As for co-maintaining: I think this is a good  idea and helps to get things going. I am currently busy with some other
packages, so I can try to help here, but in the moment would be careful with taking new responsibilities...

**Comment by carstenbauer** [comment]
> So feel free to improve this

Alright, will do. If @eschnett deems me worthy and adds me as a maintainer, I will directly modify your PR. Otherwise, I'll create a new one over the weekend.

**Comment by carstenbauer** [comment]
I implemented the changes, added some tests, updated the README and rebased. Let's wait for CI to pass.

**Review by carstenbauer — APPROVED** [review]

**Pull request merged**
